core/src/main/java/hudson/util/FileChannelWriter.java

```diff
@@ -0,0 +1,94 @@
+ * remote obviously though).
+ *
+ * <p>The goal using this is to reduce as much as we can the likeliness to see zero-length files be created in place
+ * of the original ones.</p>
+ *
+ * @see <a href="https://issues.jenkins-ci.org/browse/JENKINS-34855">JENKINS-34855</a>
+ * @see <a href="https://github.com/jenkinsci/jenkins/pull/2548">PR-2548</a>
+ */
+@Restricted(NoExternalUse.class)
+public class FileChannelWriter extends Writer {
+
+    private static final Logger LOGGER = Logger.getLogger(FileChannelWriter.class.getName());
+
+    private final Charset charset;
+    private final FileChannel channel;
+
+    /**
+     * {@link FileChannel#force(boolean)} is a <strong>very</strong> costly operation. This flag has been introduced mostly to
+     * accommodate Jenkins' previous behaviour, when using a simple {@link java.io.BufferedWriter}.
+     *
+     * <p>Basically, {@link BufferedWriter#flush()} does nothing, so when existing code was rewired to use
+     * {@link FileChannelWriter#flush()} behind {@link AtomicFileWriter} and that method actually ends up calling
+     * {@link FileChannel#force(boolean)}, many things started timing out. The main reason is probably because XStream's
+     * {@link com.thoughtworks.xstream.core.util.QuickWriter} uses <code>flush()</code> a lot.
+     * So we introduced this field to be able to still get a better integrity for the use case of {@link AtomicFileWriter}.
+     * Because from there, we make sure to call {@link #close()} from {@link AtomicFileWriter#commit()} anyway.
+     */
+    private boolean forceOnFlush;
+
+    /**
+     * See forceOnFlush. You probably never want to set forceOnClose to false.
+     */
+    private boolean forceOnClose;
+
+    /**
+     * @param filePath     the path of the file to write to.
+     * @param charset      the charset to use when writing characters.
+     * @param forceOnFlush set to true if you want {@link FileChannel#force(boolean)} to be called on {@link #flush()}.
+     * @param options      the options for opening the file.
+     * @throws IOException if something went wrong.
+     */
+    FileChannelWriter(Path filePath, Charset charset, boolean forceOnFlush, boolean forceOnClose, OpenOption... options) throws IOException {
+        this.charset = charset;
+        this.forceOnFlush = forceOnFlush;
+        this.forceOnClose = forceOnClose;
+        channel = FileChannel.open(filePath, options);
+    }
+
+    @Override
+    public void write(char cbuf[], int off, int len) throws IOException {
+        final CharBuffer charBuffer = CharBuffer.wrap(cbuf, off, len);
+        ByteBuffer byteBuffer = charset.encode(charBuffer);
+        channel.write(byteBuffer);
+    }
+
+    @Override
+    public void flush() throws IOException {
+        if (forceOnFlush) {
+            LOGGER.finest("Flush is forced");
+            channel.force(true);
+        } else {
+            LOGGER.finest("Force disabled on flush(), no-op");
```

Comment:
Should we not do *some* flush here? Otherwise turning the flag off would be result in behaviour more broken that it was before.

Reply:
@olivergondza good question.

TL;DR: `FileChannel.write()` already achieves the equivalent of `BufferedWriter.write() + BufferedWriter.flush()`.

`FileChannel`, contrary to the previous `Files.newBufferedWriter` has no Java-level buffering. So `FileChannel.write()` writes to disk, well actually sends data to the OS/disk cache depending on the system config. 

So, flushing does not make as much sense here, or at least not the same meaning as when one calls `BufferedWriter.flush()`. And BTW, even `BufferedWriter.flush()` does not actually force a full flush to disk, it's more like a cache from JVM cache/RAM to the next level, no guaranty things end up on the disk. *That* is the whole point of this change. 

`FileChannel.force()` **actually** guarantees everything is flushed to the disk. `BufferedWriter.write() + BufferedWriter.flush()` does not.

Reply:
I see. So, if I understand correctly, the data should be send to disk more eagerly than before (avoiding Java level caching) even when turned off. Not that I would object against the change, I am merely pointing out there is no way for users to switch to the old behaviour we know was working fine for ages in case this would not work as expected.

Reply:
> switch to the old behaviour we know was working fine for ages in case this would not work as expected.

`forceoflush` defaults to `false` from `AtomicFileWriter` - so we are not pushing to disk on a flush thus it should behave very similar (the data has gone from Java to the OS) - although possibly via different underlying native calls.

By default from `AtomicFileWriter` we are only making sure the data is on disk when the Writer is `close()`d  and if you want to simulate the old broken behaviour you can: see [here](https://github.com/jenkinsci/jenkins/pull/3170/files#diff-2fecb1bc52ed9dba9824e03ab7a24553R57) and [here](https://github.com/jenkinsci/jenkins/pull/3170/files#diff-2fecb1bc52ed9dba9824e03ab7a24553R146)